hashring/hashring.go

```diff
@@ -66,53 +69,46 @@ func (r *HashRing) RegisterListener(l events.EventListener) {
 	r.listeners = append(r.listeners, l)
 }
 
-func NewHashRing(hashfunc func([]byte) uint32, replicaPoints int) *HashRing {
+// New instantiates and returns a new HashRing.
+func New(hashfunc func([]byte) uint32, replicaPoints int) *HashRing {
 	ring := &HashRing{
-		hashfunc:      hashfunc,
 		replicaPoints: replicaPoints,
+		hashfunc: func(str string) int {
+			return int(hashfunc([]byte(str)))
+		},
 	}
 
 	ring.servers.byAddress = make(map[string]bool)
-	ring.servers.tree = &RBTree{}
-
+	ring.servers.tree = &RedBlackTree{}
 	return ring
 }
 
+// Checksum returns the checksum of all stored servers in the HashRing
+// Use this value to find out if the HashRing is mutated.
 func (r *HashRing) Checksum() uint32 {
 	r.servers.RLock()
-	checksum := r.servers.checksum
-	r.servers.RUnlock()
-
-	return checksum
+	defer r.servers.RUnlock()
+	return r.servers.checksum
 }
 
-// computeChecksum computes checksum of all servers in the ring
+// computeChecksum computes checksum of all servers in the ring.
 func (r *HashRing) computeChecksum() {
-	var addresses sort.StringSlice
-	var buffer bytes.Buffer
-
 	r.servers.Lock()
-	for address := range r.servers.byAddress {
-		addresses = append(addresses, address)
-	}
-
-	addresses.Sort()
-
-	for _, address := range addresses {
-		buffer.WriteString(address)
-		buffer.WriteString(";")
-	}
+	defer r.servers.Unlock()
 
+	addresses := r.getServersNoLock()
+	sort.Strings(addresses)
+	bytes := []byte(strings.Join(addresses, ";"))
 	old := r.servers.checksum
-	r.servers.checksum = farm.Fingerprint32(buffer.Bytes())
+	r.servers.checksum = farm.Fingerprint32(bytes)
```

Comment:
Do we have a test to confirm that this method is producing checksums identical to what the old one did?  Otherwise we could be in for a sad time with upgrades.